test/unit/BPool.t.sol

```diff
@@ -0,0 +1,557 @@
+
+  function test_Revert_IsBound() public view {}
+
+  function test_Revert_Finalized() public view {}
+
+  function test_Revert_MaxPoolTokens() public view {}
+
+  function test_Set_Record() public view {}
+
+  function test_Set_TokenArray() public view {}
+
+  function test_Emit_LogCall() public view {}
+
+  function test_Call_Rebind() public view {}
+}
+
+contract BPool_Unit_Rebind is Base {
+  function test_Revert_NotController() public view {}
+
+  function test_Revert_NotBound() public view {}
+
+  function test_Revert_Finalized() public view {}
+
+  function test_Revert_MinWeight() public view {}
+
+  function test_Revert_MaxWeight() public view {}
+
+  function test_Revert_MinBalance() public view {}
+
+  function test_Revert_Reentrancy() public view {}
+
+  function test_Set_TotalWeightIfDenormMoreThanOldWeight() public view {}
+
+  function test_Set_TotalWeightIfDenormLessThanOldWeight() public view {}
+
+  function test_Revert_MaxTotalWeight() public view {}
+
+  function test_Set_Denorm() public view {}
+
+  function test_Set_Balance() public view {}
+
+  function test_Pull_IfBalanceMoreThanOldBalance() public view {}
+
+  function test_Push_UnderlyingIfBalanceLessThanOldBalance() public view {}
+
+  function test_Push_FeeIfBalanceLessThanOldBalance() public view {}
+
+  function test_Emit_LogCall() public view {}
+}
+
+contract BPool_Unit_Unbind is Base {
+  function test_Revert_NotController() public view {}
+
+  function test_Revert_NotBound() public view {}
+
+  function test_Revert_Finalized() public view {}
+
+  function test_Revert_Reentrancy() public view {}
+
+  function test_Set_TotalWeight() public view {}
+
+  function test_Set_TokenArray() public view {}
+
+  function test_Set_Index() public view {}
+
+  function test_Unset_TokenArray() public view {}
+
+  function test_Unset_Record() public view {}
+
+  function test_Push_UnderlyingBalance() public view {}
+
+  function test_Push_UnderlyingFee() public view {}
+
+  function test_Emit_LogCall() public view {}
+}
+
+contract BPool_Unit_Gulp is Base {
+  function test_Revert_NotBound() public view {}
+
+  function test_Revert_Reentrancy() public view {}
+
+  function test_Set_Balance() public view {}
+
+  function test_Emit_LogCall() public view {}
+}
+
+contract BPool_Unit_GetSpotPrice is Base {
+  function test_Revert_NotBoundTokenIn() public view {}
+
+  function test_Revert_NotBoundTokenOut() public view {}
+
+  function test_Returns_SpotPrice() public view {}
+
+  function test_Revert_Reentrancy() public view {}
+}
+
+contract BPool_Unit_GetSpotPriceSansFee is Base {
+  function test_Revert_NotBoundTokenIn() public view {}
+
+  function test_Revert_NotBoundTokenOut() public view {}
+
+  function test_Returns_SpotPrice() public view {}
+
+  function test_Revert_Reentrancy() public view {}
+}
+
+contract BPool_Unit_JoinPool is Base {
+  function test_HappyPath(uint256 _poolAmountOut, uint256[8] memory _balance, uint256[8] memory _denorm) public {
```

Comment:
ideally, the test happy path should only fuzz a struct, pass the struct into a modifier, and just call the method we're testing, so for other tests we can reuse the modifier and will be very automatic.

happy path means "no revert", then for any Set, Emit, Return test,we can use the happy path.

for revert tests, we can:
'''solidity
_assumeHappyPath(struct);
struct.var = revertingValue;
_setValues(struct);
'''